generic_expression_patterns_modules/process.py

```diff
@@ -454,3 +454,256 @@ def create_all_recount2_compendium(download_dir, output_filename):
+    # The genes in `ensembl_genes` has version numbers at the end.
+    ensembl_gene_ids = pd.DataFrame(
+        data={
+            'ensembl_version': raw_ensembl_genes,
+            'ensembl_parsed': [gene_id.split('.')[0] for gene_id in raw_ensembl_genes]
+        }
+    )
+
+    # Map ensembl gene ids with version number to gene_id_mapping
+    merged_gene_id_mapping = pd.merge(
+        original_gene_id_mapping,
+        ensembl_gene_ids,
+        left_on='ensembl_gene_id',
+        right_on='ensembl_parsed',
+        how='outer'
+    )
+
+    # Set `ensembl_version` column as the index
+    merged_gene_id_mapping.set_index('ensembl_version', inplace=True)
+
+    return merged_gene_id_mapping
+
+
+def get_renamed_columns(
+        raw_ensembl_ids,
+        merged_gene_id_mapping,
+        manual_mapping,
+        DE_prior_filename
+):
+    """
+    Find the new column names and corresponding column indexes.
+
+    Arguments:
+    - raw_ensembl_ids: list of strings (ensembl gene IDs), which are columns
+                       names in raw recount2 data file;
+    - merged_gene_id_mapping: DataFrame of merged gene ID mapping;
+    - manual_mapping: dict of manual mapping (key: ensembl_id, value: gene symbol)
+
+    Returns a tuple that includes two entries. The first entry is a list
+    of hgnc gene symbols (which will be the new column names in remapped
+    recount2 data file; The second entry is a dict whose keys are hgnc gene
+    symbols and values are lists of the corresponding indexes of columns in
+    the raw recount2 data file (most lists include only one column index.)
+
+    """
+
+    updated_mapping = merged_gene_id_mapping.loc[
+        ~merged_gene_id_mapping.index.duplicated(keep="first")
+    ]
+    for ensembl_id, gene_symbol in manual_mapping.items():
+        updated_mapping.loc[ensembl_id].hgnc_symbol = gene_symbol
+
+    # Build a dict that maps hgnc symbols to column indexes in raw recount2 data
+    hgnc_to_cols = dict()
+    for col_idx, ensembl_id in enumerate(raw_ensembl_ids):
+        # Skip paralogs (whose ensembl_id includes "PAR_Y")
+        if "PAR_Y" in ensembl_id:
+            continue
+
+        hgnc_symbol = updated_mapping.loc[ensembl_id].hgnc_symbol
+
+        # Skip hgnc gene symbols that are `float` type (NaN) or empty strings
+        if type(hgnc_symbol) == float or len(hgnc_symbol) == 0:
+            continue
+
+        if hgnc_symbol in hgnc_to_cols:
+            hgnc_to_cols[hgnc_symbol].append(col_idx)
+        else:
+            hgnc_to_cols[hgnc_symbol] = [col_idx]
+
+    our_gene_ids_hgnc = list(hgnc_to_cols.keys())
+
+    published_generic_genes = get_published_generic_genes(DE_prior_filename)
+    shared_genes_hgnc = list(
+        set(our_gene_ids_hgnc).intersection(published_generic_genes)
+    )
+
+    # In Python, the order of elements in a list that is converted from a set
+    # is non-deterministic, so it is sorted here to have reproducible result.
```

Comment:
```suggestion
    # which is non-deterministic, so it is sorted here to have reproducible result.
```

Reply:
Actually, what I meant here is "the order (of ...) is non-deterministic, ...", so `which` is not needed.